app/components/FloatingAction.tsx

```diff
@@ -0,0 +1,75 @@
+import React from "react"
+import { Dimensions, View, ViewStyle } from "react-native"
+import Animated, { FadeInDown, FadeOutDown } from "react-native-reanimated"
+import { spacing } from "../theme"
+
+type Props = {
+  children: React.ReactElement
+  isScrolling?: boolean
+}
+
+enum Scroll {
+  onMomentumScrollBegin = "onMomentumScrollBegin",
+  onMomentumScrollEnd = "onMomentumScrollEnd",
+  onScrollBeginDrag = "onScrollBeginDrag",
+  onScrollEndDrag = "onScrollEndDrag",
+}
+
+export const useFloatingActionEvents = () => {
+  const [scrollState, setScrollState] = React.useState<Scroll | null>(null)
+  const [isScrolling, setIsScrolling] = React.useState(false)
+
+  const floatingScrollHandlers = {
+    [Scroll.onMomentumScrollBegin]: () => setScrollState(Scroll.onMomentumScrollBegin),
+    [Scroll.onMomentumScrollEnd]: () => setScrollState(Scroll.onMomentumScrollEnd),
+    [Scroll.onScrollBeginDrag]: () => setScrollState(Scroll.onScrollBeginDrag),
+    [Scroll.onScrollEndDrag]: () => setScrollState(Scroll.onScrollEndDrag),
+  }
+
+  React.useEffect(() => {
+    let timeout: ReturnType<typeof setTimeout>
+
+    if (scrollState === Scroll.onScrollBeginDrag) {
+      setIsScrolling(true)
+    }
+
+    if (scrollState === Scroll.onMomentumScrollBegin) {
+      clearTimeout(timeout)
+    }
+
+    if (scrollState === Scroll.onMomentumScrollEnd || scrollState === Scroll.onScrollEndDrag) {
+      timeout = setTimeout(() => {
+        setIsScrolling(false)
+      }, 500)
+    }
+
+    return () => clearTimeout(timeout)
+  }, [scrollState])
+
+  return {
+    isScrolling,
+    floatingScrollHandlers,
+  }
+}
+
+const FloatingAction = ({ children, isScrolling }: Props) => {
```

Comment:
nit: instead of calling this prop `isScrolling`, I think it would make more sense as "isVisible", since that's what it's really controlling. The floating button doesn't really care whether that state is due to scroll state or anything else.